torch/csrc/autograd/variable.h

```diff
@@ -1,150 +1,255 @@
 #pragma once
 
-// A wrapper around at::Tensor to represent autograd Variables. Variables
-// can be implicitly converted to an at::Tensor.
+#include <Python.h>
 
-#include <mutex>
-#include <memory>
-#include <vector>
-#include <functional>
-#include <ATen/ATen.h>
-
-#include "torch/csrc/assertions.h"
-#include "torch/csrc/jit/ir.h"
-#include "torch/csrc/jit/tracer_state.h"
+#include "torch/csrc/autograd/edge.h"
 #include "torch/csrc/autograd/function_hook.h"
-#include "torch/csrc/utils/auto_unique_ptr.h"
 #include "torch/csrc/autograd/variable_version.h"
-#include "torch/csrc/autograd/edge.h"
-#include "torch/csrc/Types.h"
+#include "torch/csrc/jit/tracer_state.h"
+#include "torch/csrc/utils/auto_unique_ptr.h"
+
+#include <ATen/Scalar.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Storage.h>
+#include <ATen/Tensor.h>
+#include <ATen/TensorImpl.h>
+#include <ATen/Type.h>
+
+#include <list>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
 
 namespace torch { namespace autograd {
 
-using at::Tensor;
-struct VariableImpl;
+struct Function;
+
+//===----------------------------------------------------------------------===//
+//                                Variable
+//===----------------------------------------------------------------------===//
+
+/// A `Variable` augments a `Tensor` with the ability to interact in our
+/// autograd machinery. `Variable` inherits from `Tensor` and may be converted
+/// to and from `Tensor` implicitly.
+class Variable : public at::Tensor {
+ public:
+  /// Creates a Variable that is a *view* of another (*base*) variable.
+  /// The `gradient_edge` is an optional (gradient_function, input_number) pair.
+  static Variable
+  as_view(Variable base, at::Tensor data, Edge gradient_edge = Edge());
+
+  Variable() = default;
+  Variable(at::Tensor data, bool requires_grad);
+  Variable(at::Tensor data, Edge gradient_edge);
+
+  // "Downcasts" a `Tensor` into a `Variable`. Only call this on tensors you
+  // know are Variables.
+  /*implicit*/ Variable(at::Tensor const& rhs) : at::Tensor(rhs) {}
+  /*implicit*/ Variable(at::Tensor&& rhs) noexcept
+      : at::Tensor(std::move(rhs)) {}
```

Comment:
I don't think giving up `make_variable` for an extra set of constructors is a good idea. Right now the only thing that makes constructors which take *raw tensors* and those that take *tensors assumed to be Variables* (with no checking!!) is a single `bool` argument. That's a very unsafe API. `make_variable` existed precisely for the purpose of making the distinction clear

Reply:
Ah ok. It just seemed odd to me to have to write `vector.emplace_back(make_variable(...))` instead of just `vector.emplace_back(...)`. But I get your point. `make_variable` just sounded like it should really be a constructor though. Maybe we should make them static functions and call them `Variable::wrap(...)` instead?

Reply:
The `make_` convention is not without precedent in C++. For example, you don't have a constructor for both `T*` and the perfect forwarding version of `std::shared_ptr`, you have a constructor and `std::make_shared` (in this case, you really don't want an overload because they interact poorly with perfect forwarding (see Effective Modern C++ for more details ;)). I do understand static functions are common in pure-OO codebases like Java, where top-level functions are not actually a thing, but the prevailing style in PyTorch is to prefer plain old functions over static methods.

Reply:
Yeah, I think sticking to the `make_` C++ convention is nicer than a static method

Reply:
Ok, I'll bring back the `make_*` functions.